### DISCUSSION TITLE
Using temp tables in SQL Server with raw queries and transactions?

### DISCUSSION THREAD

**Discussion by truefleet** [sequelize/sequelize#17330] [Help & Questions]
I'm writing a script to grab/transform enrollment data from a spreadsheet report and store it in a SQL Server database.  The data coming in numbers anywhere from 2500-5000 rows; I'm aware of the 1000-row limitation in inserts for SQL Server.  My current plan is to do the following:
- create a temporary table
- loop through the spreadsheet data in groups of 1000, inserting each group into the temp table
- run a MERGE statement with the temp table as source on the actual enrollments table

I'm trying to use raw query statements for this process.  Each statement is provided a sequelize transaction object, and the transaction is committed at the end.

I'm getting an error that the temporary table is an invalid object/doesn't exist when I start the second step of the process (insertion).

I am assuming that since I'm using a transaction object, that the temporary table would remain throughout the process and drop after the call to commit.  I'm wondering now if I don't fully understand how this is working/should work behind the scenes?  Or if I'm missing something that needs to be added in order to get this going?  Are temp tables even possible with sequelize and MSSQL?  I've used sequelize and transactions plenty of times in the past but this is my first time attempting DDL statements in raw queries, and with temp tables to boot.

I stripped down my initial script as a test to see if that helps me determine what's going on.  I eventually reworked the test code so that right now the first raw statement is both a CREATE and an INSERT statement, while the next statement should insert another row separately, and a final statement should count the number of entries from the previous statements.  

This is the stripped-down code; I've included part of the `new sequelize()` call from the connection service so you can see how it's created.
```
// dbConnection.service.js
const { DB_USER:username, DB_PASS:password } = process.env
const { database, ...config } = settings

let connection = new sequelize (database, username, password, {
  ...config,
  dialect: 'mssql',
  logging: console.log
})

// Added for debugging
connection.afterConnect (async (conn, config) => {
  console.log (`Database connection established`)
})

// Added for debugging
connection.afterDisconnect (async (conn) => {
  console.log (`Database connection disconnected`)
})

module.exports = connection

// ------------------------------------

// test.js
const { DBConnection } = require ('./src/services');

(async () => {
  let transaction

  try {
    transaction = await DBConnection.transaction ()

    const TEST_TABLE = 'dbo.#test_enrollments_Table'
    let cmd = `create table ${TEST_TABLE}
      (
        username varchar(10),
        catalog_Id int,
        section int
      );
      
      insert into ${TEST_TABLE} (username, catalog_Id, section) values ('testuser', 6110, 1);
    `

     // This call works
    const initialResultsObj = await DBConnection.query (cmd, { transaction })
    console.log (`initialResultsObj: ${JSON.stringify(initialResultsObj)}`)

    // This call fails
    let insertCmd = `insert into ${TEST_TABLE} (username, catalog_Id, section) values ('testuser2', 6000, 1);`
    let insertResultsObj = await DBConnection.query (insertCmd, { transaction })
    console.log (`insertResultsObj: ${JSON.stringify(insertResultsObj)}`)

    // This call is never reached
    let countCmd = `select count(*) from ${TEST_TABLE}`
    let countResultsObj = await DBConnection.query (countCmd, { transaction })
    console.log (`countResultsObj: ${JSON.stringify(countResultsObj)}`)

    await transaction.commit ()
  }

  catch (ex) {
    if (transaction)
      await transaction.rollback ()
    console.log (`\n${ex.toString()}\n`)
  }

  finally {
    console.log ('Test script complete')
    process.exit ()
  }
})()
```

When run, this is the resulting output:
```
> NODE_PATH=. node test.js

Database connection established
Database connection disconnected
Database connection established
Executing (aac99ff463cd32ffb681): BEGIN TRANSACTION;
Executing (aac99ff463cd32ffb681): create table dbo.#test_enrollments_Table
      (
        username varchar(10),
        catalog_Id int,
        section int
      );
      
      insert into dbo.#test_enrollments_Table (username, catalog_Id, section) values ('testuser', 6110, 1);
initialResultsObj: [[],1]
Executing (aac99ff463cd32ffb681): insert into dbo.#test_enrollments_Table (username, catalog_Id, section) values ('testuser2', 6000, 1);
Executing (aac99ff463cd32ffb681): ROLLBACK TRANSACTION;

SequelizeDatabaseError: Invalid object name '#test_enrollments_Table'.

Test script complete
```

I see that the db connection is made and dropped before being made again, but I've determined that happens in any other project I ran a test on so I'm not focused on that at the moment.  You'll see that the table is created and one row is inserted, but then the next statement, that is separate from the first two, fails because the table doesn't exist.

I have tried adding the schema name to the table variable, and also dropping the variable name entirely and hard-coding the table name for each statement, all with the same results.  I have also tried creating the temp table directly from a model and creating it using `sync()` but the same thing happens.

I'm not sure what else I might be missing.  Any guidance for getting the temp table to be recognized throughout the transaction would be appreciated.

**Comment by carlaulloa** [comment]
Did you solve it?

**Reply by truefleet** [reply]
Unfortunately no, not as I had hoped.  For the purposes of this project I was able to work around it by simply removing the temp designation on the table, so each time the script runs it creates a normal table to do the 1000-row insertions.  Then I have to manually clean up and call 'DROP dbo.<new_table_name>' at the end of the script, which is kind of what I wanted to avoid in the first place.

I'm refactoring another project and it looks like I might have another use-case for this process, so I'm still actively looking for a solution.
